tclreadlineCompleter.tcl

```diff
@@ -3469,76 +4202,153 @@ namespace eval tclreadline {
+    ####################################################################
+
     proc complete(unknown) {text start end line pos mod} {
-        switch -- $pos {
-            1       { return [CompleteFromList $text [CommandCompletion $text]] }
+        switch -- ${pos} {
+            1 {
+                return [CompleteFromList ${text} [CommandCompletion ${text}]]
+            }
             default { return [DisplayHints ?arg?] }
         }
         return ""
     }
 
+
+    ####################################################################
+    ####################################################################
+    ## unload ?switches? fileName
+    ## unload ?switches? fileName packageName
+    ## unload ?switches? fileName packageName interp
+    ####################################################################
+    ####################################################################
+    proc complete(unload) {text start end line pos mod} {
+    
+        set opts [RemoveUsedOptions $line {-nocomplain -keeplibrary -- <fileName>} {--}]
+        set res [string trim [TryFromList $text $opts]]
+        set prev [PreviousWord $start $line]
+        if {"--" == $prev} {
+            return ""
+        } else {
+            return $res
+        }
+
+        return ""
+    }
+
+    ####################################################################
+    ####################################################################
+    ## unset ?-nocomplain? ?--? ?name name name ...?
+    ####################################################################
+    ####################################################################
     proc complete(unset) {text start end line pos mod} {
-        return [VarCompletion $text]
+            switch -- ${pos} {
+                1 {
+                    return [CompleteFromList ${text} {-- -nocomplain}]
+                }
+                2 {
+                    switch -- [PreviousWord ${start} ${line}] {
+                        -nocomplain {
+                            return [CompleteFromList ${text} --]
+                        }
+                        -- {
+                            return [VarCompletion ${mod}]
+                        }
+                        default { 
+                            return [DisplayHints <?name?>] 
+                        }
+                    }
+                }
+                default {
+                    return [VarCompletion ${text}]
+                        }
+                }
     }
 
+
+    ####################################################################
+    ####################################################################
+    ## update ?idletasks?
+    ####################################################################
+    ####################################################################
     proc complete(update) {text start end line pos mod} {
-        switch -- $pos {
+        switch -- ${pos} {
```

Comment:
The braces around the variable name are unnecessary here since there is no parsing ambiguity without them.